src/org/exist/xquery/functions/fn/FunUnparsedText.java

```diff
@@ -0,0 +1,149 @@
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.regex.Pattern;
+
+import static org.exist.xquery.FunctionDSL.*;
+
+public class FunUnparsedText extends BasicFunction {
+
+    private final static String LINE_REGEX = "\\r\\n|\\r|\\n";
+
+    private final static FunctionParameterSequenceType PARAM_HREF = optParam("href", Type.STRING, "the URI to load text from");
+    private final static FunctionParameterSequenceType PARAM_ENCODING = param("encoding", Type.STRING, "character encoding of the resource");
+
+    static final FunctionSignature [] FS_UNPARSED_TEXT = functionSignatures(
+        new QName("unparsed-text", Function.BUILTIN_FUNCTION_NS),
+        "reads an external resource (for example, a file) and returns a string representation of the resource",
+        returnsOpt(Type.STRING),
+        arities(
+                arity(PARAM_HREF),
+                arity(PARAM_HREF, PARAM_ENCODING)
+        ));
+
+    static final FunctionSignature[] FS_UNPARSED_TEXT_LINES = functionSignatures(
+        new QName("unparsed-text-lines", Function.BUILTIN_FUNCTION_NS),
+        "reads an external resource (for example, a file) and returns its contents as a sequence of strings, one for each line of text in the string representation of the resource",
+        returnsOptMany(Type.STRING),
+        arities(
+                arity(PARAM_HREF),
+                arity(PARAM_HREF, PARAM_ENCODING)
+        ));
+
+    static final FunctionSignature [] FS_UNPARSED_TEXT_AVAILABLE = functionSignatures(
+        new QName("unparsed-text-available", Function.BUILTIN_FUNCTION_NS),
+        "determines whether a call on the fn:unparsed-text function with identical arguments would return a string",
+        returnsOpt(Type.STRING),
+        arities(
+                arity(PARAM_HREF),
+                arity(PARAM_HREF, PARAM_ENCODING)
+        ));
+
+    public FunUnparsedText(final XQueryContext context, final FunctionSignature signature) {
+        super(context, signature);
+    }
+
+    @Override
+    public Sequence eval(Sequence[] args, Sequence contextSequence) throws XPathException {
+        final String encoding = args.length == 2 ? args[1].getStringValue() : null;
+        if (!args[0].isEmpty()) {
+            final String content;
+            try {
+                content = readContent(args[0].getStringValue(), encoding);
+            } catch (XPathException e) {
+                if (isCalledAs("unparsed-text-available")) {
+                    return BooleanValue.FALSE;
+                }
+                throw e;
+            }
+            if (isCalledAs("unparsed-text-available")) {
+                return BooleanValue.TRUE;
+            }
+            if (isCalledAs("unparsed-text-lines")) {
+                final Pattern pattern = PatternFactory.getInstance().getPattern(LINE_REGEX);
```

Comment:
For large text files, it would be much more efficient to use `LineNumberReader` on the `InputStream` than to read the file into a string, split it into many strings (two copies of the data), and then return a `StringValue`.